File: src/lib/huffman/decoder.ts
import type { HuffmanNode, DecodingResult, DecodingStatistics } from './types';
import { deserializeTree, isLeaf } from './tree';

const MAGIC_BYTES = new Uint8Array([0x48, 0x55, 0x46, 0x46]);

export function decode(data: Uint8Array): DecodingResult {
  if (data.length < 14) {
    throw new Error('Invalid Huffman file: too short');
  }

  const magic = data.slice(0, 4);
  if (!magic.every((byte, i) => byte === MAGIC_BYTES[i])) {
    throw new Error('Invalid Huffman file: wrong magic bytes');
  }

  const version = data[4];
  if (version !== 1) {
    throw new Error(`Unsupported Huffman file version: ${version}`);
  }

  const padding = data[5];
  const view = new DataView(data.buffer, data.byteOffset);
  const originalLength = view.getUint32(6, false);
  const treeLength = view.getUint32(10, false);

  if (originalLength === 0) {
    return {
      decodedText: '',
      statistics: {
        encodedSize: data.length,
        encodedBits: data.length * 8,
        decodedSize: 0,
        decodedBits: 0,
        compressionRatio: 0,
        totalChars: 0,
      },
    };
  }

  const treeData = new TextDecoder().decode(data.slice(14, 14 + treeLength));
  const tree = deserializeTree(treeData);

  if (!tree) {
    throw new Error('Invalid Huffman file: failed to deserialize tree');
  }

  const encodedData = data.slice(14 + treeLength);
  const decodedText = decodeBytes(encodedData, tree, padding, originalLength);

  const encodedBits = data.length * 8;
  const decodedBits = decodedText.length * 8;

  const statistics: DecodingStatistics = {
    encodedSize: data.length,
    encodedBits,
    decodedSize: decodedText.length,
    decodedBits,
    compressionRatio: decodedBits > 0 ? Math.round((encodedBits / decodedBits) * 10000) / 100 : 0,
    totalChars: decodedText.length,
  };

  return { decodedText, statistics };
}

function decodeBytes(
  encodedData: Uint8Array,
  tree: HuffmanNode,
  padding: number,
  originalLength: number
): string {
  let binaryString = '';
  for (const byte of encodedData) {
    binaryString += byte.toString(2).padStart(8, '0');
  }

  if (padding > 0) {
    binaryString = binaryString.slice(0, -padding);
  }

  const result: string[] = [];
  let currentNode: HuffmanNode = tree;
  let charsDecoded = 0;

  for (const bit of binaryString) {
    if (charsDecoded >= originalLength) {
      break;
    }

    if (bit === '0') {
      currentNode = currentNode.left!;
    } else {
      currentNode = currentNode.right!;
    }

    if (!currentNode) {
      throw new Error('Invalid encoded data: tree traversal failed');
    }

    if (isLeaf(currentNode)) {
      if (currentNode.char !== null) {
        result.push(currentNode.char);
        charsDecoded++;
      }
      currentNode = tree;
    }
  }

  return result.join('');
}

File: backend/huffman/decoder.py
import json
import struct
from .tree import HuffmanNode


class HuffmanDecoder:
    MAGIC_BYTES = b'HUFF'

    def __init__(self):
        self.tree: HuffmanNode | None = None
        self.original_length: int = 0
        self.padding: int = 0

    def decode(self, data: bytes) -> str:
        if len(data) < 14:
            raise ValueError("Invalid Huffman file: too short")

        magic = data[:4]
        if magic != self.MAGIC_BYTES:
            raise ValueError("Invalid Huffman file: wrong magic bytes")

        version = data[4]
        if version != 1:
            raise ValueError(f"Unsupported Huffman file version: {version}")

        self.padding = data[5]
        self.original_length = struct.unpack('>I', data[6:10])[0]
        tree_length = struct.unpack('>I', data[10:14])[0]

        if self.original_length == 0:
            return ''

        tree_json = data[14:14 + tree_length].decode('utf-8')
        tree_data = json.loads(tree_json)

        if not tree_data:
            return ''

        self.tree = HuffmanNode.from_dict(tree_data)

        encoded_data = data[14 + tree_length:]
        return self._decode_bytes(encoded_data)

    def _decode_bytes(self, encoded_data: bytes) -> str:
        if not self.tree:
            return ''

        binary_string = ''.join(format(byte, '08b') for byte in encoded_data)

        if self.padding > 0:
            binary_string = binary_string[:-self.padding]

        result = []
        current_node = self.tree
        chars_decoded = 0

        for bit in binary_string:
            if chars_decoded >= self.original_length:
                break

            if bit == '0':
                current_node = current_node.left
            else:
                current_node = current_node.right

            if current_node is None:
                raise ValueError("Invalid encoded data: tree traversal failed")

            if current_node.is_leaf():
                if current_node.char is not None:
                    result.append(current_node.char)
                    chars_decoded += 1
                current_node = self.tree

        return ''.join(result)

    def get_statistics(self, encoded_data: bytes, decoded_text: str) -> dict:
        encoded_bits = len(encoded_data) * 8
        original_bits = len(decoded_text) * 8

        return {
            'encodedSize': len(encoded_data),
            'encodedBits': encoded_bits,
            'decodedSize': len(decoded_text),
            'decodedBits': original_bits,
            'compressionRatio': round(encoded_bits / original_bits * 100, 2) if original_bits > 0 else 0,
            'totalChars': len(decoded_text)
        }
